nncf/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,631 @@
+                continue
+            for end_node in act_output_shape[shape]:
+                if end_node.main_id - start_node.main_id > max_block_size:
+                    continue
+                if end_node.node_type in IGNORED_NAME_OPERATORS:
+                    continue
+                if end_node.main_id <= start_node.main_id:
+                    continue
+                is_one_edge = (end_node.main_id - start_node.bottom_id) == 1
+
+                # CHECK RULES
+                all_rules_is_true = True
+                for rule_fn in fn_rules:
+                    if not rule_fn(sgraph, start_node, end_node):
+                        all_rules_is_true = False
+                        break
+                if all_rules_is_true and not is_one_edge:
+                    blocks.append(BuildingBlock(start_node, end_node))
+
+    sorted_blocks = sorted(blocks, key=cmp_to_key(compare_for_building_block))
+    if not allow_linear_combination:
+        sorted_blocks = remove_linear_combination(sorted_blocks)
+    if not allow_nested_blocks:
+        sorted_blocks = remove_nested_blocks(sorted_blocks)
+    building_blocks_in_orig_graph = restore_node_name_in_orig_graph(sorted_blocks, orig_graph)
+
+    return building_blocks_in_orig_graph
+
+
+def remove_nested_blocks(sorted_blocks: List[BuildingBlock]) -> List[BuildingBlock]:
+    """
+    Remove nested building blocks.
+    """
+    return [list(group_block)[-1] for _, group_block in groupby(sorted_blocks, lambda block: block.start_node.main_id)]
+
+def get_group_of_dependent_blocks(blocks: List[BuildingBlock]) -> Dict[int, int]:
+    """
+    Building blocks can be categorized into groups. Blocks that follow each other in the graph
+    (that is, they are connected by one edge) belong to the same group.
+    """
+    groups = {}
+    idx = 0
+    groups = { idx: [] }
+    for i in range(len(blocks) - 1):
+        start_node_key_i1 = blocks[i + 1][0]
+        end_node_key_i  = blocks[i][1]
+        if start_node_key_i1 == end_node_key_i:
+            groups[idx].append(i)
+        else:
+            groups[idx].append(i)
+            idx +=1
+            groups[idx] = []
+    groups[idx].append(len(blocks) - 1)
+
+    return groups
+
+def get_building_blocks_info(bblocks: List[BuildingBlock], compressed_model: NNCFNetwork) -> List[BuildingBlockInfo]:
+    bblocks_info = []
+    for block in bblocks:
+        op_addresses = get_all_node_op_addresses_in_block(compressed_model, block)
+        modules = get_all_modules_in_blocks(compressed_model, op_addresses)
+        block_type = get_type_building_block(op_addresses)
+        bblocks_info.append(BuildingBlockInfo(block, op_addresses, modules, block_type))
+    return bblocks_info
+
+def get_all_node_op_addresses_in_block(compressed_model: NNCFNetwork, block: BuildingBlock) -> Set[OperationAddress]:
+    """
+    Returns set of operation addresses of all layers included in the block.
```

Comment:
Isn't there already a function in `NNCFNetwork` that solves the same problem?

Reply:
No, there isn`t.